### ISSUE TITLE
Pattern and Additional properties

### ISSUE THREAD

**Issue by delvedor** [mcollina/fast-json-stringify#5]
Below the proposal to how implement pattern and additional properties.

#### patternProperties:
Must be an object, where the key is a valid regex and the value is an object.
Example:
```javascript
"patternProperties": {
    "^S_": { "type": "string" },
    "^I_": { "type": "integer" }
  }
```
`patternProperties` will work only for the properties that are not explicitly listed in the properties object.

#### additionalProperties:
Must be a boolean or an object.
If the is setted to false and there are properties that are not explicitly listed in the properties object we should throw an error.
If `additionalProperties` is used in conjunction with `patternProperties` the pecking order will be:
`properties -> patternProperties -> additionalProperties`
Example: 
```javascript
"additionalProperties": {
  "type": "string"
}
```

#### Discussion:
The json schema is very bad documented, so before start to implement these functionalities we need to discuss about few things.
- **additionalProperties**
  If is setted to `{ type: 'string' }` and any additional property is not a string we have two possible cases:
  1. throw an error
  2. stringify the property as string

- **patternProperties**
  Basically the same as above, if a property matches to a pattern with `{ type: 'number' }` but the property is not a number we have two cases:
  1. throw an error
  2. stringify the property as number

If I forgot or I didn't see some edge case please write it below.

**Comment by mcollina** [comment]
Use http://npm.im/fast-safe-stringify to stringify the property, we can even be gentle and not have a strong type check for now, we can add that later (maybe).

**Comment by delvedor** [comment]
Ok but we should stringify it as is or as the pattern/additional property says?

Example:
*schema*
```javascript
properties: {...}
additionalProperties: {
  type: 'string'
}
```
*object to stringify*
```javascript
{ addProp: 42 } // addProp is a number and not a string
```
The result will be
```javascript
'{"addProp":42}'   // original type
                   // or 
'{"addProp":"42"}' // additionaProperty wanted type
```

**Comment by mcollina** [comment]
We should try to coerce and serialize. If not possible, throw.
Il giorno lun 19 set 2016 alle 14:02 Tomas Della Vedova <
notifications@github.com> ha scritto:

**Comment by delvedor** [comment]
Edge case:
Schema:
```javascript
{
  title: 'title',
  type: 'object',
  properties: {...}
  patternProperties: {
    foo: {
      type: 'string'
    }
  }
}
```
if the user tries to stringify `objfoo: {a: true}`what we should do?
1) Stringify as `[object Object]`
2) Stringify with `fast-safe-stringify`
3) throw

I think the best option is the first one, because the user must know what is passing to `fast-json-stringify`, and if he sets a patternProperty, he's expecting the all the matching property have a specific type.
Furthermore I guess that `[object Object]` is faster than de others (I didn't do any bench).

In the other cases, if `foo: { type: 'number' } -stringify-as-> NaN`
If `{ type: 'boolean' } -stringify-as-> true`

**Comment by mcollina** [comment]
IMHO `[object Object]` sounds good. In fact, I would call `toString()` on it to coerce.

```
> ({}).toString()
'[object Object]'
```

**Comment by delvedor** [comment]
If `additionalProperties` is setted to *false* and there's a property that does not match to `properties` and `patternProperties`, what behavior we want to adopt?
We can throw an error or "discard" the property when is time to stringify it.

I think is better to "discard" the property, but I'm not sure about this because the documentation is very unclear.

Documentation:
- [json-schema](http://json-schema.org/latest/json-schema-validation.html#anchor64)
- [understanding-json-schema](https://spacetelescope.github.io/understanding-json-schema/reference/object.html#properties)

**Comment by mcollina** [comment]
@delvedor yes, discard the property. It should behave like a "filter".

**Comment by delvedor** [comment]
Both successfully implemented, closing.
